impl/src/main/java/impl/UserResourceImpl.java

```diff
@@ -71,16 +80,47 @@ public UserResourceImpl(UserDao userDao, ResponseHeaderHolder responseHeaderHold
     public Observable<User> generateToken(@NotNull String openIdToken) {
         return openIdValidator.validate(openIdToken).flatMap(validOpenId ->
                 userDao.getUserByEmail(validOpenId.email)
-                .onErrorResumeNext(t -> error(new WebException(HttpResponseStatus.INTERNAL_SERVER_ERROR, "failed.to.search.for.user", t)))
+                .onErrorResumeNext(t -> error(new WebException(HttpResponseStatus.INTERNAL_SERVER_ERROR, FAILED_TO_SEARCH_FOR_USER, t)))
                 .switchIfEmpty(addUserToDatabase(validOpenId.name, validOpenId.email, validOpenId.picture))
-                .map(user -> {
-                            ApplicationToken applicationToken = applicationTokenCreator.createApplicationToken(validOpenId, user.getId());
-                            addAsCookie(applicationToken, user);
-                            return user;
-                        }
-                ));
+                .flatMap(user -> updateUserNameAndPicture(user, validOpenId))
+                .map(user -> addApplicationTokenToHeader(validOpenId, user)));
+    }
+
+    private User addApplicationTokenToHeader(ImmutableOpenIdToken validOpenId, User user) {
+        ApplicationToken applicationToken = applicationTokenCreator.createApplicationToken(validOpenId, user.getId());
+        addAsCookie(applicationToken, user);
+        return user;
     }
 
+    private Observable<User> updateUserNameAndPicture(User user, ImmutableOpenIdToken openId) {
+        if(!hasUpdatedNameOrPicture(user, openId)) {
+            return just(user);
+        }
+
+        return userDao.updateUser(user.getId(), openId.name, openId.picture)
```

Comment:
Is there always a proper picture returned here even if the user never has set a profile pic?

Reply:
In the case of Google, yes. But maybe not with other vendors, need to make sure we can default to a picture in that case so that we dont need to check for null values anywhere else

Reply:
I think that for this MVP. we should scope that out, because in the real world. the ui only supports google. But i added a default image, just to make sure that the application works for a user without a picture. The image is not hosted by rocket fuel as you can see. But creating a image and making it work through webpack feels a bit out of scope, with the mvp in mind.